pinot-core/src/main/java/org/apache/pinot/core/plan/AggregationPlanNode.java

```diff
@@ -154,4 +118,181 @@ private static boolean isFitForDictionaryBasedPlan(AggregationFunction[] aggrega
+    // TODO: Use the same operator for both of them so that COUNT(*), MAX(col) can be optimized
+    if (filterOperator.isResultMatchingAll() && !hasFilteredPredicates) {
+      if (isFitForMetadataBasedPlan(aggregationFunctions)) {
+        return Pair.of(null, new MetadataBasedAggregationOperator(aggregationFunctions,
+            _indexSegment.getSegmentMetadata(), Collections.emptyMap()));
+      } else if (isFitForDictionaryBasedPlan(aggregationFunctions, _indexSegment)) {
+        Map<String, Dictionary> dictionaryMap = new HashMap<>();
+        for (AggregationFunction aggregationFunction : aggregationFunctions) {
+          String column = ((ExpressionContext) aggregationFunction.getInputExpressions().get(0)).getIdentifier();
+          dictionaryMap.computeIfAbsent(column, k -> _indexSegment.getDataSource(k).getDictionary());
+        }
+        return Pair.of(null,
+            new DictionaryBasedAggregationOperator(aggregationFunctions, dictionaryMap,
+            numTotalDocs));
+      }
+    }
+
+    if (starTrees != null && !StarTreeUtils.isStarTreeDisabled(_queryContext)) {
+      // Use star-tree to solve the query if possible
+
+      AggregationFunctionColumnPair[] aggregationFunctionColumnPairs =
+          StarTreeUtils.extractAggregationFunctionPairs(aggregationFunctions);
+      if (aggregationFunctionColumnPairs != null) {
+        Map<String, List<CompositePredicateEvaluator>> predicateEvaluatorsMap =
+            StarTreeUtils.extractPredicateEvaluatorsMap(_indexSegment, _queryContext.getFilter(),
+                filterPlanNode.getPredicateEvaluatorMap());
+        if (predicateEvaluatorsMap != null) {
+          for (StarTreeV2 starTreeV2 : starTrees) {
+            if (StarTreeUtils.isFitForStarTree(starTreeV2.getMetadata(), aggregationFunctionColumnPairs, null,
+                predicateEvaluatorsMap.keySet())) {
+
+              TransformOperator transformOperator = new StarTreeTransformPlanNode(starTreeV2,
+                  aggregationFunctionColumnPairs, null,
+                      predicateEvaluatorsMap, _queryContext.getDebugOptions()).run();
+              AggregationOperator aggregationOperator = new AggregationOperator(aggregationFunctions,
+                  transformOperator, numTotalDocs, true);
+
+              return Pair.of(transformOperator, aggregationOperator);
+            }
+          }
+        }
+      }
+    }
+
+    TransformOperator transformOperator = new TransformPlanNode(_indexSegment, _queryContext,
+          expressionsToTransform, DocIdSetPlanNode.MAX_DOC_PER_CALL, filterOperator).run();
+    AggregationOperator aggregationOperator = new AggregationOperator(aggregationFunctions,
+        transformOperator, numTotalDocs, false);
+
+    return Pair.of(transformOperator, aggregationOperator);
+  }
+
+  /**
+   * Builds the operator to be used for non filtered aggregations
+   */
+  private BaseOperator<IntermediateResultsBlock> buildNonFilteredAggOperator() {
```

Comment:
What I meant is to move the current code into this method, and implement `buildFilteredAggOperator()` separately. The reason being:
1. The metadata/dictionary based operator and star-tree does not apply to the filtered aggregation
2. Sharing `buildOperators()` method can bring extra overhead to non-filtered aggregations

Reply:
Fixed